extensions/access_log_policy/plugin.cc

```diff
@@ -0,0 +1,143 @@
+    ::Envoy::Extensions::Common::Wasm::Null::NullPluginRootRegistry;
+using google::protobuf::util::JsonParseOptions;
+using google::protobuf::util::Status;
+
+NULL_PLUGIN_ROOT_REGISTRY;
+
+#endif
+
+namespace {
+
+bool setFilterStateValue(bool log) {
+  auto r = setFilterStateStringValue(::Wasm::Common::kAccessLogPolicyKey,
+                                     log ? "yes" : "no");
+  if (r != WasmResult::Ok) {
+    logWarn(toString(r));
+    return false;
+  }
+  return true;
+}
+
+}  // namespace
+
+constexpr long int kDefaultLogWindowDurationNanoseconds =
+    43200000000000;  // 12h
+
+constexpr StringView kClientIp = "client_ip";
+constexpr StringView kPrincipal = "principal";
+constexpr StringView kDestination = "destination";
+constexpr StringView kAddress = "address";
+constexpr StringView kConnection = "connection";
+constexpr StringView kUriSanPeerCertificate = "uri_san_peer_certificate";
+
+static RegisterContextFactory register_AccessLogPolicy(
+    CONTEXT_FACTORY(PluginContext), ROOT_FACTORY(PluginRootContext));
+
+bool PluginRootContext::onConfigure(size_t) {
+  if (::Wasm::Common::TrafficDirection::Outbound ==
+      ::Wasm::Common::getTrafficDirection()) {
+    throw EnvoyException("ASM Acess Logging Policy is an inbound filter only.");
+  }
+  WasmDataPtr configuration = getConfiguration();
+  JsonParseOptions json_options;
+  Status status =
+      JsonStringToMessage(configuration->toString(), &config_, json_options);
+  if (status != Status::OK) {
+    logWarn("Cannot parse Stackdriver plugin configuration JSON string " +
+            configuration->toString() + ", " + status.message().ToString());
+    return false;
+  }
+
+  if (config_.has_log_window_duration()) {
+    log_time_duration_nanos_ =
+        ::google::protobuf::util::TimeUtil::DurationToNanoseconds(
+            config_.log_window_duration());
+  } else {
+    log_time_duration_nanos_ = kDefaultLogWindowDurationNanoseconds;
+  }
+
+  // We currently cache the key based on client_ip and principal only.
+  IstioDimensions::InitIstioDimensions({kClientIp, kPrincipal});
+
+  return true;
+}
+
+void PluginContext::onLog() {
+  // Check if request is a failure.
+  int64_t response_code = 0;
+  getValue({"response", "code"}, &response_code);
+  // If request is a failure, log it.
+  if (response_code != 200) {
+    setFilterStateValue(true);
+    return;
+  }
+
+  // If request is not a failure, check cache to see if it should be logged or
+  // not, based on last time a successful request was logged for this client ip
+  // and principal combination.
+  std::string downstream_ip = "";
+  getStringValue({kDestination, kAddress}, &downstream_ip);
+  std::string source_principal = "";
+  getStringValue({kConnection, kUriSanPeerCertificate}, &source_principal);
+  IstioDimensions istio_dimension;
+  istio_dimension.map(kClientIp, downstream_ip);
+  istio_dimension.map(kPrincipal, source_principal);
+
+  long int last_log_time_nanos = lastLogTimeNanos(istio_dimension);
```

Comment:
Is there a reason why you use architecture dependent type? Is this intentional to be sensitive to 32/64-bits?

Reply:
not really.. used what we are using in stackdriver extension for edge reporting... changing everything to be absl::time and duration